### ISSUE TITLE
jackc/pgx does not  send query to the postgres sometimes. and returns empty rows in response for such type of queries.

### ISSUE THREAD

**Issue by ashishmittal13** [jackc/pgx#1129]
Postgress instance is running on AWS.

**Comment by Marahin** [comment]
I'm facing the same issue, code that sometimes returns 0 (and doesn't actually end up shooting a query):

```go
func GetSpots() ([]*Spot, error) {
	spots := make([]*Spot, 0)

	rows, err := db.Conn.Query(context.Background(), "SELECT * FROM web_spot")
	if err != nil {
		return spots, err
	}

	for rows.Next() {
		row := &Spot{}

		rows.Scan(&row.ID, &row.Name, &row.CreatedAt)

		spots = append(spots, row)
	}

	return spots, nil
}
```

First call is always a hit: 

```
db_1   | 2021-12-20 22:46:30.646 UTC [36] LOG:  execute lrupsc_1_0: SELECT * FROM web_spot
```

And then all subsequent calls are just my warning messages:
```
bot_1  | 2021/12/20 22:46:33 stderr: time="2021-12-20T22:46:33Z" level=info msg=InteractionCreate name=book
bot_1  | 2021/12/20 22:46:33 stderr: time="2021-12-20T22:46:33Z" level=warning msg="GetSpots() returned 0"
bot_1  | 2021/12/20 22:46:43 stderr: time="2021-12-20T22:46:43Z" level=info msg=InteractionCreate name=book
bot_1  | 2021/12/20 22:46:43 stderr: time="2021-12-20T22:46:43Z" level=warning msg="GetSpots() returned 0"
bot_1  | 2021/12/20 22:46:46 stderr: time="2021-12-20T22:46:46Z" level=info msg=InteractionCreate name=book
bot_1  | 2021/12/20 22:46:46 stderr: time="2021-12-20T22:46:46Z" level=warning msg="GetSpots() returned 0"
bot_1  | 2021/12/20 22:46:54 stderr: time="2021-12-20T22:46:54Z" level=info msg=InteractionCreate name=book
bot_1  | 2021/12/20 22:46:54 stderr: time="2021-12-20T22:46:54Z" level=warning msg="GetSpots() returned 0"
```

While the record is still in the database, intact, and no queries were shot by pgx.

**Comment by jackc** [comment]
@Marahin I don't see any error checks after the `rows.Next()` loop. You always need to check `rows.Err()`.

**Comment by ashishmittal13** [comment]
@Marahin  Exactly Same thing is happening with me. 

@jackc  Even if their is no check for rows.Err() , in ideal world there should not be error.  But anyways,  i am checking it`s value and it is nil.

**Comment by cristaloleg** [comment]
@Marahin looks like there is no `defer rows.Close()`  so after reading data connection is not released so all subsequent queries will be blocked.

**Comment by ashishmittal13** [comment]
Hey @cristaloleg  Do we also need to do something similar during the write calls. 
like defer batchResults.Close() ?

**Comment by cristaloleg** [comment]
@ashishmittal13 yeah, after every query (read/write) you need to release the connection. AFAIR it's mentioned in readme or docs.

**Comment by jackc** [comment]
By chance is some proxy or load balancer being used? #1130 describes what appears to be the same problem, but only when connecting through haproxy. A direct connection to the database works.

**Comment by G2G2G2G** [comment]
@cristaloleg https://github.com/jackc/pgx/blob/master/examples/todo/main.go#L73 why does the example not close?

**Comment by jackc** [comment]
> https://github.com/jackc/pgx/blob/master/examples/todo/main.go#L73 why does the example not close?

rows are automatically closed by `Next()` when no more rows are available. Also automatically closed if a `Scan()` returns an error. See https://pkg.go.dev/github.com/jackc/pgx/v4#Rows and https://pkg.go.dev/github.com/jackc/pgx/v4#hdr-Query_Interface.

**Comment by G2G2G2G** [comment]
So what cristaloleg said above is wrong since next closes them? That's what I thought. Thanks for the links.

**Comment by ashishmittal13** [comment]
So this is not due to not what @cristaloleg  said.

**Comment by cristaloleg** [comment]
Yep, looks I was looking at the wrong api.

**Comment by G2G2G2G** [comment]
I haven't been able to get this error, but the other guy did.. perhaps it's latency related? I'm right by the datacenter so not really any latency.

**Comment by ashishmittal13** [comment]
@Marahin Are you still seeing this issue?
@G2G2G2G  I don`t think it  is related to any latency issue.

**Comment by Marahin** [comment]
@ashishmittal13 in the mean time I managed to change to `db/sql` and _also_ applied your fixes. If I find enough time to try pgx again, I'll let you know if the issue is fixed.

**Comment by bfontaine** [comment]
Anything new on this?
